src/nginx/t/rollout_config_download_metadata.t

```diff
@@ -0,0 +1,311 @@
+is( $r->{uri}, '/v1/services/endpoints-test.cloudendpointsapis.com:check',
+  ':check was called');
+is( $r->{headers}->{host}, "127.0.0.1:${ServiceControlPort}",
+  'Host header was set');
+is( $r->{headers}->{'content-type'}, 'application/x-protobuf',
+  ':check Content-Type was protocol buffer');
+
+# report
+$r = shift @requests;
+
+is( $r->{verb}, 'POST', ':report verb was post' );
+is( $r->{uri}, '/v1/services/endpoints-test.cloudendpointsapis.com:report',
+  ':report was called');
+is( $r->{headers}->{host}, "127.0.0.1:${ServiceControlPort}",
+  'Host header was set');
+is( $r->{headers}->{'content-type'}, 'application/x-protobuf',
+  ':check Content-Type was protocol buffer' );
+
+################################################################################
+
+sub bookstore {
+  my ( $t, $port, $file ) = @_;
+  my $server = HttpServer->new( $port, $t->testdir() . '/' . $file )
+    or die "Can't create test server socket: $!\n";
+  local $SIG{PIPE} = 'IGNORE';
+
+  $server->on( 'GET', '/shelves?key=this-is-an-api-key', <<'EOF');
+HTTP/1.1 200 OK
+Connection: close
+
+{ "shelves": [
+    { "name": "shelves/1", "theme": "Fiction" },
+    { "name": "shelves/2", "theme": "Fantasy" }
+  ]
+}
+EOF
+  $server->run();
+}
+
+my @quota_responses      = ();
+my $quota_response_index = 0;
+
+sub servicecontrol {
+  my ( $t, $port, $file, $done) = @_;
+  my $server = HttpServer->new( $port, $t->testdir() . '/' . $file )
+    or die "Can't create test server socket: $!\n";
+  local $SIG{PIPE} = 'IGNORE';
+
+  $server->on( 'POST',
+    '/v1/services/endpoints-test.cloudendpointsapis.com:check', <<'EOF');
+HTTP/1.1 200 OK
+Connection: close
+
+EOF
+
+  $server->on('POST',
+    '/v1/services/endpoints-test.cloudendpointsapis.com:allocateQuota',
+    <<'EOF');
+HTTP/1.1 200 OK
+Connection: close
+
+EOF
+
+  $server->on_sub('POST', 
+    '/v1/services/endpoints-test.cloudendpointsapis.com:report', sub {
+    my ($headers, $body, $client) = @_;
+    print $client <<'EOF';
+HTTP/1.1 200 OK
+Connection: close
+
+EOF
+
+    $t->write_file($done, ':report done');
+  });
+
+  $server->run();
+}
+
+sub servicemanagement {
+  my ( $t, $port, $file) = @_;
+  my $server = HttpServer->new( $port, $t->testdir() . '/' . $file )
+    or die "Can't create test server socket: $!\n";
+  local $SIG{PIPE} = 'IGNORE';
+
+  $server->on_sub('GET', 
+  '/v1/services/endpoints-test.cloudendpointsapis.com/configs/2017-05-01r0', sub {
+    my ($headers, $body, $client) = @_;
+
+    my $resp_body = <<'EOF';
+{
+  "name": "endpoints-test.cloudendpointsapis.com",
+  "title": "Bookstore",
+  "http": {
+    "rules": [
+      {
+        "selector": "EchoGetMessage",
+        "get": "/shelves"
+      }
+    ]
+  },
+  "usage": {
+    "rules": [
+      {
+        "selector": "EchoGetMessage",
+        "allowUnregisteredCalls": true
+      }
+    ]
+  },
+  "control": {
+EOF
+    $resp_body .= '  "environment": "http://127.0.0.1:' . 
+        $ServiceControlPort . '"' ."\n";
```

Comment:
can this service config use

" control: {
    "environment":  "http://127.0.0.1:$ServiceControlPort"
}"

It will be easier to read.  

BTW,  it is better put this service config at the top file, next to nginx config, and pass it to serviemanagemnt server.